src/coreclr/vm/comdelegate.cpp

```diff
@@ -801,49 +898,67 @@ static Stub* SetupShuffleThunk(MethodTable* pDelMT, MethodDesc* pTargetMeth)
     }
     CONTRACTL_END;
 
-    GCX_PREEMP();
-
+    bool isInstRetBuff = (!pTargetMeth->IsStatic() && pTargetMeth->HasRetBuffArg() && IsRetBuffPassedAsFirstArg());
     DelegateEEClass * pClass = (DelegateEEClass *)pDelMT->GetClass();
 
+    // Look for a thunk cached on the delegate class first. Note we need a different thunk for instance methods with a
+    // hidden return buffer argument because the extra argument switches place with the target when coming from the caller.
+    Stub* pShuffleThunk = isInstRetBuff ? pClass->m_pInstRetBuffCallStub : pClass->m_pStaticCallStub;
+    if (pShuffleThunk)
+        return pShuffleThunk->GetEntryPoint();
+
+    GCX_PREEMP();
+
     MethodDesc *pMD = pClass->GetInvokeMethod();
 
+    // We haven't already setup a shuffle thunk, go do it now (which will cache the result automatically).
     StackSArray<ShuffleEntry> rShuffleEntryArray;
-    GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray);
+    bool isSimple = GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray);
+    if (isSimple)
+    {
+        ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
 
-    ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
+        LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
+        if (pLoaderAllocator->IsCollectible())
+        {
+            pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
+        }
 
-    LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
-    if (pLoaderAllocator->IsCollectible())
+        pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
+    }
+    else
     {
-        pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
+#if defined(TARGET_RISCV64)
+        pShuffleThunk = CreateILDelegateShuffleThunk(pMD, isInstRetBuff);
+#else
+        _ASSERTE(FALSE);
+        return NULL;
+#endif // TARGET_RISCV64
     }
 
-    Stub* pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
     if (!pShuffleThunk)
     {
         COMPlusThrowOM();
     }
 
-    if (!pTargetMeth->IsStatic() && pTargetMeth->HasRetBuffArg() && IsRetBuffPassedAsFirstArg())
+    // Cache the shuffle thunk
+    Stub** ppThunk = isInstRetBuff ? &pClass->m_pInstRetBuffCallStub : &pClass->m_pStaticCallStub;
+    Stub* pExistingThunk = InterlockedCompareExchangeT(ppThunk, pShuffleThunk, NULL);
+    if (pExistingThunk != NULL)
     {
-        if (InterlockedCompareExchangeT(&pClass->m_pInstRetBuffCallStub, pShuffleThunk, NULL ) != NULL)
+        if (isSimple)
```

Comment:
Should the same check be in https://github.com/dotnet/runtime/blob/770df102b0dfbf16c22ef64a270a989495cfe145/src/coreclr/vm/class.cpp#L135-L145 as well?

You should be able to use `Stub::HasExternalEntryPoint()` to tell the kind of stub you are dealing with.

Reply:
I think so. BTW, any reason StubLinkStubManager::RemoveStubRange is called only on m_pStaticCallStub and not m_pInstRetBuffCallStub? These two are created the same way.